Mathlib/RingTheory/Flat/CategoryTheory.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.RingTheory.Flat.Basic
+import Mathlib.Algebra.Category.ModuleCat.Monoidal.Symmetric
+import Mathlib.Algebra.Category.ModuleCat.Projective
+import Mathlib.Algebra.Category.ModuleCat.Abelian
+import Mathlib.CategoryTheory.Monoidal.Tor
+
+/-!
+# Tensoring with a flat module is an exact functor
+
+In this file we prove that tensoring with a flat module is an exact functor.
+
+## Main results
+- `Module.Flat.iff_tensorLeft_preservesFiniteLimits`: an `R`-module `M` is flat if and only if
+  left tensoring with `M` preserves finite limits, i.e. the functor `- ⊗ M` is left exact.
+
+- `Module.Flat.iff_tensorRight_preservesFiniteLimits`: an `R`-module `M` is flat if and only if
+  right tensoring with `M` preserves finite limits, i.e. the functor `M ⊗ -` is left exact.
+
+- `Module.Flat.higherTorIsoZero`: if an `R`-module `M` is flat, then `Torⁿ(M, N) ≅ 0` for all `N`
+  and all `n ≥ 1`.
+
+## TODO
+
+- Prove that vanishing `Tor`-groups implies flat.
+
+-/
+
+universe u
+
+open CategoryTheory MonoidalCategory Abelian
+
+namespace Module.Flat
+
+variable {R : Type u} [CommRing R] (M : ModuleCat.{u} R)
+
+open scoped MonoidalCategory in
+set_option maxHeartbeats 400000 in
+-- In two goals, we need to use `simpa` in one; and `simp` in the other.
+set_option linter.unnecessarySimpa false in
+noncomputable instance [flat : Flat R M] {X Y : ModuleCat.{u} R} (f : X ⟶ Y) :
+    Limits.PreservesLimit (Limits.parallelPair f 0) (tensorLeft M) where
+  preserves {c} hc := by
+    let ι : c.pt ⟶ X := c.π.app .zero
+    have mono0 : Mono ι :=
+      { right_cancellation := fun {Z} g h H => by
+          let c' : Limits.Cone (Limits.parallelPair f 0) :=
+          ⟨Z, ⟨fun | .zero => h ≫ ι | .one => 0, by rintro _ _ (⟨j⟩|⟨j⟩) <;> simpa [ι] using H⟩⟩
+
+          rw [hc.uniq c' g, hc.uniq c' h] <;>
+          rintro (⟨j⟩|⟨j⟩) <;> simpa [ι] using H }
+    have exact0 : Exact ι f := by
```

Comment:
Could you use https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Homology/ShortComplex/Exact.html instead? (It is likely that the notion from `Algebra.Homology.Exact` will be removed eventually. Actually, your PR reminds me that I should remove it...)

Reply:
I am not very proficient in the homological algebra part of library. Can you check if you like the changes now please?

Reply:
The PR has gotten quite big after the change, if necessary I will split it.